### PR TITLE
[otbn,doc] Documentation improvements

### PR DESCRIPTION
**Pull request by ctopal**
1st commit updates block diagram to seperate OTBN block diagram and `otbn_core` block diagram.
2nd commit adds a README to help document OTBN ISS.

### CODE REVIEW

**Comment by ctopal** [comment]
It's not as much in detail as I wanted to but I aimed to give some sort of general idea about what is where regarding the internals of the ISS. Let me know if you think I missed some parts.

File: hw/ip/otbn/doc/otbn_blockarch.svg [hunk 1]
```diff
@@ -915,1041 +2056,519 @@
...
+       y="44.941654"
+       style="stroke-width:1.92105055"><tspan
+   style="font-weight:bold"
+   id="tspan3811">O</tspan>pen<tspan
+   style="font-weight:bold"
+   id="tspan3813">T</tspan>itan <tspan
+   style="font-weight:bold"
+   id="tspan3809">B</tspan>ig <tspan
+   style="font-weight:bold"
+   id="tspan3807">N</tspan>umber Module - OTBN</tspan></text>
```
Comment (andreaskurth) [inline hw/ip/otbn/doc/otbn_blockarch.svg:2516]:
```suggestion
   id="tspan3807">N</tspan>umber Accelerator - OTBN</tspan></text>
```

File: hw/ip/otbn/doc/otbn_blockarch.svg [hunk 2]
```diff
@@ -8,13 +8,13 @@
    xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
    xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
    version="1.1"
-   viewBox="0 0 949.50442 767.90681"
+   viewBox="0 0 1076.6736 886.2356"
```
Comment (andreaskurth) [inline hw/ip/otbn/doc/otbn_blockarch.svg:11]: (This comment applies to this file but not exactly to this line.)

I think the lines for the randomness returned by EDN and the sideloaded key _outside_ `otbn_core` should be colored the same as inside (green if I'm not mistaken)

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# OTBNSim - OTBN Simulation Software
+
+`OTBNSim` class represents the ISS in a traditional sense.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:3]: What does 'traditional' mean here? Can you elaborate?

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+It has methods for starting the executing, cycle-accurate simulation tracing and also abilty to simulate injected errors and wiping as a response to those errors.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:4]:
```suggestion
It has methods for starting the execution, for cycle-accurate simulation tracing, and for injecting errors and react to those errors with a secure wipe.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+As explained in the overall [OTBN DV documentation](../../doc/dv/_index.md), this simulator is driven essentially from the SystemVerilog model of the OTBN module.
+
+Simulator works in a step-by-step fashion and also has multiple different methods to mimic external stimulus to OTBN.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:7]:
```suggestion
The simulator works in a step-by-step fashion and it has multiple methods to apply external stimuli to OTBN.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+At a typical run without errors ISS does the following:
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:8]:
```suggestion
In a typical run without errors, the ISS does the following:
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`/
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:9]:
```suggestion
 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+ 2. Loading decoded program to a local list in `sim.py`.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:10]:
```suggestion
 2. Load the decoded program to a local list in `sim.py`.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+ 3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
+ 4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.
+
+## Co-Simulation with RTL
+Co-simulation works by using a tracer module (`otbn_tracer`) that spits out some logs from the RTL and having a serialised result in a text form on `stdout` to produce same style of logs from the Python model.
+These two seperate logs generated by the model and the RTL gets compared for checking correct behaviour.
+For more information about how OTBN RTL produces traces see. [Tracer README](../tracer/README.md).
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:17]:
```suggestion
For more information about how OTBN RTL produces traces see the [Tracer README](../tracer/README.md).
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+Simulator works in a step-by-step fashion and also has multiple different methods to mimic external stimulus to OTBN.
+At a typical run without errors ISS does the following:
+ 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`/
+ 2. Loading decoded program to a local list in `sim.py`.
+ 3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
+ 4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.
+
+## Co-Simulation with RTL
+Co-simulation works by using a tracer module (`otbn_tracer`) that spits out some logs from the RTL and having a serialised result in a text form on `stdout` to produce same style of logs from the Python model.
+These two seperate logs generated by the model and the RTL gets compared for checking correct behaviour.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:16]:
```suggestion
These two separate logs generated by the model and the RTL get compared.
If the state of the RTL ever differs from the state of the model, an error is reported.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+Simulator works in a step-by-step fashion and also has multiple different methods to mimic external stimulus to OTBN.
+At a typical run without errors ISS does the following:
+ 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`/
+ 2. Loading decoded program to a local list in `sim.py`.
+ 3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
+ 4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.
+
+## Co-Simulation with RTL
+Co-simulation works by using a tracer module (`otbn_tracer`) that spits out some logs from the RTL and having a serialised result in a text form on `stdout` to produce same style of logs from the Python model.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/otbnsim/README.md:15]:
```suggestion
For co-simulation of RTL and ISS, the `otbn_tracer` module logs state changes of the RTL, and the ISS logs state changes of the Python model.
```
I don't think the `stdout` mechanics should be in the first sentence explaining co-simulation, so I removed it. If you think it should be mentioned at this point, I suggest adding a second sentence.

**Review by andreaskurth — COMMENTED** [review]
This is very useful, thanks @ctopal!

I've added a few suggestions that would make the text more clear for me. What concerns missing content, maybe someone who has not worked with OTBN ISS before could read this, then get a walk through from you in the call tomorrow, and afterwards add whatever is missing to this doc?

**Comment by ctopal** [comment]
Thanks @andreaskurth for the feedback! I applied them and extended the tracing related part a little bit.

**Review by andreaskurth — APPROVED** [review]
LGTM, thanks @ctopal!

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# OTBNSim - OTBN Simulation Software
+
+`OTBNSim` class represents the simulator that runs a model of the OTBN block as a whole.
+Therefore it includes abstract models for the the state of the core, memory modules of it and also the registers.
+It has methods for starting the execution, for cycle-accurate simulation tracing, and for injecting errors and react to those errors with a secure wipe.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:5]:
```suggestion
It has methods for starting the execution, cycle-accurate simulation tracing, and for injecting errors and reacting to those errors with a secure wipe.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
+# OTBNSim - OTBN Simulation Software
+
+`OTBNSim` class represents the simulator that runs a model of the OTBN block as a whole.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:3]:
```suggestion
`OTBNSim` class represents a simulator that models of the whole OTBN block.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+Therefore it includes abstract models for the the state of the core, memory modules of it and also the registers.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:4]:
```suggestion
Therefore it includes abstract models for the the state of the core, its memory modules, and also the register file.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+It has methods for starting the execution, for cycle-accurate simulation tracing, and for injecting errors and react to those errors with a secure wipe.
+As explained in the overall [OTBN DV documentation](../../doc/dv/_index.md), this simulator is driven essentially from the SystemVerilog model of the OTBN module.
+
+The simulator works in a step-by-step fashion and it has multiple methods to apply external stimuli to OTBN.
+In a typical run without errors, the ISS does the following:
+ 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:10]:
```suggestion
 1. Decode the program on the path given by `iss_wrapper.cc` by using the `decode_file` method in `decode.py`.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+ 2. Load the decoded program to a local list in `sim.py`.
+ 3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:12]:
```suggestion
 3. With each `step` command from the SystemVerilog side, update the simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+ 4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:13]:
```suggestion
 4. Once the step is done, pass the generated trace to `iss_wrapper.cc`, which then passes it on to `OTBNTraceChecker`.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+## Co-Simulation with RTL
+For co-simulation of RTL and ISS, the `otbn_tracer` module logs state changes of the RTL, and the ISS logs state changes of the Python model.
+Trace entries from the simulated core (aka. from RTL) appear as a result of DPI callbacks while ISS trace entries appear in the trace checker through `ISSWrapper` using `OnIssTrace` method after sending a step command to `OTBNSim`.
+These two separate logs generated by the model and the RTL get compared for checking correct behaviour.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:18]:
```suggestion
 To check for correct behaviour, the two separate logs generated by the model and the RTL are compared.
```

File: hw/ip/otbn/dv/otbnsim/README.md [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+For more information about how OTBN RTL produces traces see the [Tracer README](../tracer/README.md).
+To see C++ program that compares both traces check `otbn_trace_checker.cc` on `../model/otbn_trace_entry`.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/dv/otbnsim/README.md:20]:
```suggestion
To see the C++ program that compares both traces, check the method `otbn_trace_checker.cc` in `../model/otbn_trace_entry`.
```

**Review by hcallahan-lowrisc — APPROVED** [review]
Looks good! Thanks for the explanations @ctopal . I've left a few small suggestions for grammar changes, but the overall flow is great and the information is very useful.

**Comment by andreaskurth** [comment]
This changes only the documentation, so we don't need CI to pass (other than linting maybe, but the Quick Lint check already passed). Merging

**Pull request merged**